xl-objects-multipart.go

```diff
@@ -222,22 +222,20 @@ func (xl xlObjects) CompleteMultipartUpload(bucket string, object string, upload
 		return "", toObjectErr(err, bucket, multipartObjFile)
 	}
 
-	// Attempt a rename of the upload id to temporary location, if
-	// successful then delete it.
+	// Delete the incomplete file place holder.
 	uploadIDPath := path.Join(mpartMetaPrefix, bucket, object, uploadID, incompleteFile)
-	tempUploadIDPath := path.Join(tmpMetaPrefix, bucket, object, uploadID, incompleteFile)
-	if err = xl.storage.RenameFile(minioMetaBucket, uploadIDPath, minioMetaBucket, tempUploadIDPath); err == nil {
-		if err = xl.storage.DeleteFile(minioMetaBucket, tempUploadIDPath); err != nil {
-			return "", toObjectErr(err, minioMetaBucket, tempUploadIDPath)
-		}
-		return s3MD5, nil
-	}
-	// Rename if failed attempt to delete the original file.
 	err = xl.storage.DeleteFile(minioMetaBucket, uploadIDPath)
 	if err != nil {
 		return "", toObjectErr(err, minioMetaBucket, uploadIDPath)
 	}
 
+	// Validate if there are other incomplete upload-id's present for
+	// the object, if yes do not attempt to delete 'uploads.json'.
+	if entries, err := xl.storage.ListDir(minioMetaBucket, path.Join(mpartMetaPrefix, bucket, object)); err == nil {
```

Comment:
need to hold a lock because just after ListDir someone might have done initiate-new-multipart-upload